src/metabase/email.clj

```diff
@@ -129,6 +131,74 @@
       (log/warn e (trs "Failed to send email"))
       {::error e})))
 
+(defn- reconfigure-retrying "No-op. Will be overridden." [_ _])
+
+(defsetting email-retry-max-attempts
+  (deferred-tru "The maximum number of attempts for delivering a single email.")
+  :type :integer
+  :default (run-mode-val
+            (:dev :test) 1
+            5)
+  :on-change #'reconfigure-retrying)
+
+(defsetting email-retry-initial-interval
+  (deferred-tru "The initial retry delay in milliseconds when delivering emails.")
+  :type :integer
+  :default (run-mode-val
+             (:dev :test) 1
+             15000)
+  :on-change #'reconfigure-retrying)
+
+(defsetting email-retry-multiplier
+  (deferred-tru "The delay multiplier between attempts to deliver a single email.")
+  :type :double
+  :default 2.0
+  :on-change #'reconfigure-retrying)
+
+(defsetting email-retry-randomizaion-factor
+  (deferred-tru "The randomization factor of the retry delay when delivering emails.")
+  :type :double
+  :default 0.1
+  :on-change #'reconfigure-retrying)
+
+(defsetting email-retry-max-interval-millis
+  (deferred-tru "The maximum delay beetween attempts to deliver a single email.")
+  :type :integer
+  :default 120000
+  :on-change #'reconfigure-retrying)
+
+(defn- make-retry-state
+  "Returns an email sender wrapping [[send-message!]] retrying according to the
+  email retry configuration."
+  []
+  (let [retry (retry/random-exponential-backoff-retry
+               "send-email-retry"
+               {:max-attempts (email-retry-max-attempts)
+                :initial-interval-millis (email-retry-initial-interval)
+                :multiplier (email-retry-multiplier)
+                :randomization-factor (email-retry-randomizaion-factor)
+                :max-interval-millis (email-retry-max-interval-millis)
+                :retry-on-result-pred #(and (map? %) (instance? Throwable (::error %)))})]
+    {:retry retry
```

Comment:
I don't see retry getting used anywhere (besides in sender below), so is it needed in this map? 

Hmm, it could be helpful for debugging though. Is that right?

Reply:
> I don't see retry getting used anywhere (besides in sender below), so is it needed in this map?

Right, it's not currently used, I just thought that if we don't create one on every call, then why not keep a reachable reference to it.  (There is another [discussion](https://github.com/metabase/metabase/pull/21881#issuecomment-1111195584) about the retry object.)